pkg/process/monitor/process_monitor.go

```diff
@@ -0,0 +1,285 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux
+// +build linux
+
+package monitor
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"runtime"
+	"sync"
+
+	"github.com/vishvananda/netlink"
+
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/gopsutil/process"
+)
+
+const (
+	processMonitorMaxEvents = 2048
+)
+
+var (
+	processMonitorLock sync.RWMutex
+	processMonitor     *ProcessMonitor
+)
+
+type ProcessMonitor struct {
+	m  sync.Mutex
+	wg sync.WaitGroup
+
+	initializing bool
+
+	events chan netlink.ProcEvent
+	done   chan struct{}
+	errors chan error
+
+	procEventCallbacks map[ProcessEventType][]*ProcessCallback
+	callbackRunner     chan func()
+	callbackRunnerDone chan struct{}
+}
+
+type ProcessEventType int
+
+const (
+	EXEC ProcessEventType = iota
+	FORK
+	EXIT
+)
+
+type ProcessMetadataField int
+
+const (
+	ANY ProcessMetadataField = iota
+	NAME
+	MAPFILE
+)
+
+type ProcessCallback struct {
+	Event    ProcessEventType
+	Metadata ProcessMetadataField
+	Regex    *regexp.Regexp
+	Callback func(pid uint32)
+}
+
+func (pm *ProcessMonitor) enqueueCallback(callback *ProcessCallback, pid uint32) {
+	pm.callbackRunner <- func() { callback.Callback(pid) }
+}
+
+// Initialize() will scan all running processes and execute matching callbacks
+// Once it's done all new events from netlink socket will be processed by the main async loop
+func (pm *ProcessMonitor) Initialize() error {
+	fn := func(pid int) error {
+		for _, c := range pm.procEventCallbacks[EXEC] {
+			pm.evalEXECCallback(c, uint32(pid))
+		}
+		return nil
+	}
+
+	pm.m.Lock()
+	defer pm.m.Unlock()
+	if err := util.WithAllProcs(util.HostProc(), fn); err != nil {
+		return fmt.Errorf("process monitor init, scanning all process failed %s", err)
+	}
+	// enable events to be processed
+	pm.initializing = false
```

Comment:
that is a bit counter intuitive, maybe change the field to `isInitialized` and set it to `true` here?